### PR TITLE
gpuarray GpuGemmBatch op

### PR DESCRIPTION
**Pull request by cooijmanstim**
Add GpuGemmBatch op in the gpuarray backend. Includes optimizations and tests. Depends on Theano/libgpuarray#142, solves #3946.

Please double-check the `perform` implementation.

### CODE REVIEW

**Comment by cooijmanstim** [comment]
Is there an easy way to include negative tests with `makeTester`? I'd like to make sure it crashes properly on bad shapes and types.

**Comment by nouiz** [comment]
The elemwise makeTester allow to test negative case, but I didn't saw it in this version.

File: theano/sandbox/gpuarray/blas.py [hunk 1]
```diff
@@ -331,6 +331,92 @@ def c_code_cache_version(self):
...
+        return Apply(self, [C, alpha, A, B, beta], [C.type()])
+
+    def perform(self, node, inputs, outputs):
+        C, alpha, A, B, beta = inputs
+        if self.inplace and C.flags.forc:
+            C *= beta
+            C += alpha * blas.batched_dot(A, B)
+            outputs[0][0] = C
+        else:
+            outputs[0][0] = alpha * blas.batched_dot(A, B) + beta * C
```
Comment (abergeron) [inline theano/sandbox/gpuarray/blas.py]: Where do you take you batched_dot implementation?  I didn't see anything in the libgpuarray PR.

You can just remove the perform if it doesn't work.

File: theano/sandbox/gpuarray/blas.py [hunk 2]
```diff
@@ -331,6 +331,92 @@ def c_code_cache_version(self):
...
+        }
+        """ % vars
+        if config.gpuarray.sync:
+            code += """
+            GpuArray_sync(&%(out)s->ga);
+            """ % vars
+        return code
+
+    def c_code_cache_version(self):
+        return None
```
Comment (abergeron) [inline theano/sandbox/gpuarray/blas.py]: This should return a tuple with some number in order to be actually cached.

File: theano/sandbox/gpuarray/opt.py [hunk 3]
```diff
@@ -743,6 +743,14 @@ def local_gpua_gemm(node, context_name):
 
 
 @register_opt('fast_compile')
+@op_lifter([tensor.blas.BatchedDot])
+def local_gpua_gemmbatch(node, context_name):
+    a, b = node.inputs
+    c = tensor.zeros((a.shape[0], a.shape[1], b.shape[2]))
```
Comment (abergeron) [inline theano/sandbox/gpuarray/opt.py]: Make that an AllocEmpty since you set beta to 0.

**Comment by abergeron** [comment]
You also have these flake8 errors:

```
/home/travis/build/Theano/Theano/theano/sandbox/gpuarray/blas.py:334:1: E302 expected 2 blank lines, found 1
/home/travis/build/Theano/Theano/theano/tensor/blas.py:2441:1: E302 expected 2 blank lines, found 1
```

File: theano/sandbox/gpuarray/blas.py (hunk 1 above)
Comment (cooijmanstim) [inline theano/sandbox/gpuarray/blas.py]: I've removed it. I was confused about what it's for and I meant to use tensor.blas.batched_dot which this PR also adds.

**Pull request merged**